%\VignetteIndexEntry{AffymetrixPreprocessCN Vignette for Affymetrix}
%\VignetteDepends{crlmm, genomewidesnp6Crlmm, cacheSweave, ff}
%\VignetteKeywords{crlmm, SNP 6, copy number, SNP}
%\VignettePackage{crlmm}
\documentclass{article}
\usepackage{graphicx}
\usepackage{natbib}
\usepackage{url}
\newcommand{\Rfunction}[1]{{\texttt{#1}}}
\newcommand{\Rmethod}[1]{{\texttt{#1}}}
\newcommand{\Rcode}[1]{{\texttt{#1}}}
\newcommand{\Robject}[1]{{\texttt{#1}}}
\newcommand{\Rpackage}[1]{{\textsf{#1}}}
\newcommand{\Rclass}[1]{{\textit{#1}}}
\newcommand{\oligo}{\Rpackage{oligo }}
\newcommand{\R}{\textsf{R}}
\newcommand{\crlmm}{\Rpackage{crlmm}}
\usepackage[margin=1in]{geometry}

\begin{document}
\title{Preprocessing \& Genotyping Affymetrix Arrays for Copy Number Analysis}
\date{\today}
\author{Rob Scharpf}
\maketitle

<<setup, echo=FALSE, results=hide>>=
options(continue=" ", width=70)
@

%\section{Estimating copy number}

%At present, software for copy number estimation is provided only for the
%Affymetrix 6.0 platform.

\begin{abstract}

  This vignette describes the setup needed to analyze Affymetrix 6.0
  (or 5.0) CEL files and the steps for preprocessing and
  genotyping. These steps must be completed prior to copy number
  analyses in \crlmm{}.  After completing these steps, users can refer
  to the \verb+copynumber+ vignette.

\end{abstract}

\section{Set up}

<<cdfname, results=hide>>=
library(crlmm)
library(ff)
library(cacheSweave)
@

This vignette analyzes HapMap samples assayed on the Affymetrix 6.0
platform.  The annotation package for this platform is
\Rpackage{genomewidesnp6Crlmm}.  We assign the name of the annotation
package without the \verb+Crlmm+ postfix to the name \verb+cdfName+.

<<cdfname>>=
cdfName <- "genomewidesnp6"
@


The HapMap CEL files are stored in a local directory assigned to
\verb+pathToCels+ in the following code. The genotyping step will
create several files with \verb+ff+ extensions. We will store these
files to the path indicated by \verb+outdir+.

<<setup>>=
pathToCels <- "/thumper/ctsa/snpmicroarray/hapmap/raw/affy/1m"
if(getRversion() < "2.13.0"){
	rpath <- getRversion()
} else rpath <- "trunk"
outdir <- paste("/thumper/ctsa/snpmicroarray/rs/ProcessedData/crlmm/", rpath, "/copynumber_vignette", sep="")
dir.create(outdir, recursive=TRUE, showWarnings=FALSE)
@

By providing the path in \verb+outdir+ as an argument to the \R{}
function \Rfunction{ldPath}, all of the \verb+ff+ files created during
the genotyping step will be stored in \verb+outdir+.

<<ldpath>>=
ldPath(outdir)
@

This vignette uses the \R{} package \Rpackage{cacheSweave} to cache
long computations.  The following step is only necessarily if one
wishes to cache some of the computations.  In particular, we specify
that the cached computations will be saved in the \verb+outdir+
through the function \Rfunction{setCacheDir}.  Users should refer to
the \Rpackage{cacheSweave} package for additional details regarding
cacheing.

<<cachedir, echo=FALSE>>=
setCacheDir(outdir)
@

The \R{} functions \Rfunction{ocProbesets} and \Rfunction{ocSamples}
manage the RAM required for our analysis. See the documentation for
these functions and the \verb+CopyNumberOverview+ vignette for
additional details.

<<ram>>=
ocProbesets(100000)
ocSamples(200)
@


Next we indicate the local directory that contains the CEL files. For
the purposes of this vignette, we only analyze the CEPH ('C') and
Yoruban ('Y') samples.

<<celfiles>>=
celFiles <- list.celfiles(pathToCels, full.names=TRUE, pattern=".CEL")
celFiles <- celFiles[substr(basename(celFiles), 13, 13) %in% c("C", "Y")]
@

Finally, copy number analyses using \crlmm{} require specification of
a batch variable that is used to indicate which samples were processed
together.  For example, if some of the samples were processed in April
and another set of samples were processed in June, we could name the
batches 'April' and 'June', respectively.  A useful surrogate for
batch is often the chemistry plate or the scan date of the array. For
the HapMap CEL files analyzed in this vignette, the CEPH (C) and
Yoruban (Y) samples were prepared on separate chemistry plates.  In
the following code chunk, we extract the population identifier from
the CEL file names and assign these identifiers to the variable
\Robject{plate}.

<<plates>>=
plates <- substr(basename(celFiles), 13, 13)
@

\section{Preprocessing and genotyping.}

The preprocessing steps for copy number estimation includes quantile
normalization of the raw intensities for each probe and a step that
summarizes the intensities of multiple probes at a single locus.  For
example, the Affymetrix 6.0 platform has 3 or 4 identical probes at
each polymorphic locus and the normalized intensities are summarized
by a median.  For the nonpolymorphic markers on Affymetrix 6.0, only
one probe per locus is available and the summarization step is not
needed.  After preprocessing the arrays, the \crlmm{} package
estimates the genotype using the CRLMM algorithm and provides a
confidence score for the genotype calls.  The function
\Rfunction{genotype} performs both the preprocessing and genotyping.

<<LDS_genotype, cache=TRUE>>=
cnSet <- genotype(celFiles, batch=plates, cdfName=cdfName)
@

The value returned by genotype is an instance of the class
\Robject{CNSet}.  The normalized intensities, genotype calls, and
confidence scores are stored as \verb+ff+ objects in the
\verb+assayData+ slot.  A concise summary of this object can be
obtained throught the \Rfunction{print} or \Rfunction{show} methods.

<<show>>=
print(cnSet)
@

Note that the object is fairly small as the intensities and genotype
calls are stored on disk rather than in active memory.

<<objectsize>>=
object.size(cnSet)
@

We save the \Robject{cnSet} object in a local directory for subsequent
copy number analysis in the \verb+copynumber+ vignette.

<<save,cache=TRUE>>=
saveObject <- function(outdir, cnSet) {
	save(cnSet, file=file.path(outdir, "cnSet.rda"))
	TRUE
}
(cnset.saved <- saveObject(outdir, cnSet))
@

Users can proceed to the \verb+copynumber+ vignette for copy number
analyses.



\section{Session information}
<<sessionInfo, results=tex>>=
toLatex(sessionInfo())
@


\end{document}
