stupidSplineBasis <- function(x,knots){
  x <- pmin(x,knots[3])
  x <- pmax(x,knots[1])
  cbind(1, x, x^2, pmax(0, (x-knots[2]))^2)
}

changeToCrlmmAnnotationName <- function(x){
  pkgDir <- system.file(package=THISPKG)
  wanted <- paste(tolower(gsub("_", "", x)), "crlmm.stuff", sep=".")
  file.path(pkgDir, "extdata", wanted)
}

getCrlmmAnnotationName <- function(x){
  paste(tolower(gsub("_", "", x)), "Crlmm", sep="")
}

medianSummaries <- function(mat, grps)
  .Call("R_subColSummarize_median", mat, grps, PACKAGE = "preprocessCore")

intMedianSummaries <- function(mat, grps)
  as.integer(medianSummaries(mat, grps))



## .crlmmPkgEnv is an enviroment that will
## store all the variables used by the pkg.
## it's meant to not overwrite user's variables
## and get rid of the NOTES generated by
## R CMD check

isLoaded <- function(dataset, environ=.crlmmPkgEnv)
  exists(dataset, envir=environ)
getVarInEnv <- function(dataset, environ=.crlmmPkgEnv){
  if (!isLoaded(dataset))
    stop("Variable ", dataset, " not found in .crlmmPkgEnv")
  environ[[dataset]]
}

list2SnpSet <- function(x, returnParams=FALSE){
  pd <- data.frame(SNR=x[["SNR"]], gender=x[["gender"]],
                   batchQC=rep(x[["batchQC"]], ncol(x[["calls"]])),
                   row.names=colnames(x[["calls"]]))
  pdv <- data.frame(labelDescription=c("Signal-to-noise Ratio",
                      "Gender: Male (1) and Female (2)",
                      "Quality score for batch"),
                    row.names=c("SNR", "gender", "batchQC"))

  recall <- length(x[["DD"]]) > 1
  if (returnParams){
    if (recall){
      fd <- data.frame(SNPQC=x[["SNPQC"]],
                       cAA=x[["params"]][["centers"]][,1],
                       cAB=x[["params"]][["centers"]][,2],
                       cBB=x[["params"]][["centers"]][,3],
                       sAA=x[["params"]][["scales"]][,1],
                       sAB=x[["params"]][["scales"]][,2],
                       sBB=x[["params"]][["scales"]][,3],
                       nAA=x[["params"]][["N"]][,1],
                       nAB=x[["params"]][["N"]][,2],
                       nBB=x[["params"]][["N"]][,3],
                       spAA=x[["DD"]][,1],
                       spAB=x[["DD"]][,2],
                       spBB=x[["DD"]][,3],
                       row.names=rownames(x[["calls"]]))
      fdv <- data.frame(labelDescription=c(
                          "SNP Quality Score",
                          "Center AA", "Center AB", "Center BB",
                          "Scale AA", "Scale AB", "Scale BB",
                          "N AA", "N AB", "N BB",
                          "Shift in parameters AA",
                          "Shift in parameters AB",
                          "Shift in parameters BB"),
                        row.names=c(
                          "SNPQC",
                          "cAA", "cAB", "cBB",
                          "sAA", "sAB", "sBB",
                          "nAA", "nAB", "nBB",
                          "spAA", "spAB", "spBB"))
    }else{
      fd <- data.frame(SNPQC=x[["SNPQC"]],
                       cAA=x[["params"]][["centers"]][,1],
                       cAB=x[["params"]][["centers"]][,2],
                       cBB=x[["params"]][["centers"]][,3],
                       sAA=x[["params"]][["scales"]][,1],
                       sAB=x[["params"]][["scales"]][,2],
                       sBB=x[["params"]][["scales"]][,3],
                       nAA=x[["params"]][["N"]][,1],
                       nAB=x[["params"]][["N"]][,2],
                       nBB=x[["params"]][["N"]][,3],
                       row.names=rownames(x[["calls"]]))
      fdv <- data.frame(labelDescription=c(
                          "SNP Quality Score",
                          "Center AA", "Center AB", "Center BB",
                          "Scale AA", "Scale AB", "Scale BB",
                          "N AA", "N AB", "N BB"),
#                          "Shift in parameters AA",
#                          "Shift in parameters AB",
#                          "Shift in parameters BB"),
                        row.names=c(
                          "SNPQC",
                          "cAA", "cAB", "cBB",
                          "sAA", "sAB", "sBB",
                          "nAA", "nAB", "nBB"))
    }
  }else{
    if (recall){
      fd <- data.frame(SNPQC=x[["SNPQC"]],
                       spAA=x[["DD"]][,1],
                       spAB=x[["DD"]][,2],
                       spBB=x[["DD"]][,3],
                       row.names=rownames(x[["calls"]]))
      fdv <- data.frame(labelDescription=c("SNP Quality Score",
                          "Shift in parameters AA",
                          "Shift in parameters AB",
                          "Shift in parameters BB"),
                        row.names=c("SNPQC", "spAA", "spAB", "spBB"))
    }else{
      fd <- data.frame(SNPQC=x[["SNPQC"]],
                       row.names=rownames(x[["calls"]]))
      fdv <- data.frame(labelDescription=c("SNP Quality Score"),
                        row.names=c("SNPQC"))
    }
  }
  new("SnpSet",
      assayData=assayDataNew("lockedEnvironment",
        call=x[["calls"]], callProbability=x[["confs"]]),
      phenoData=new("AnnotatedDataFrame",
        data=pd, varMetadata=pdv),
      featureData=new("AnnotatedDataFrame",
        data=fd, varMetadata=fdv),
      annotation=x[["pkgname"]])
}

loader <- function(theFile, envir, pkgname){
	theFile <- file.path(system.file(package=pkgname),
			     "extdata", theFile)
	if (!file.exists(theFile))
		stop("File ", theFile, " does not exist in ", pkgname)
	load(theFile, envir=envir)
}

celDates <- function(celfiles){
	if(!all(file.exists(celfiles))) stop("1 or more cel file does not exist")
	celdates <- vector("character", length(celfiles))
	celtimes <- vector("character", length(celfiles))
	for(i in seq(along=celfiles)){
		if(i %% 100 == 0) cat(".")
		tmp <- read.celfile.header(celfiles[i], info="full")$DatHeader
		tmp <- strsplit(tmp, "\ +")
		celdates[i] <- tmp[[1]][6]
		celtimes[i] <- tmp[[1]][7]
	}
	tmp <- paste(celdates, celtimes)
	celdts <- strptime(tmp, "%m/%d/%y %H:%M:%S")
	return(celdts)
}

isPackageLoaded <- function(pkg){
	stopifnot(is.character(pkg))
	pkg <- paste("package:", pkg, sep="")
	pkg %in% search()
}

validCdfNames <- function(){
	c("genomewidesnp6",
	  "genomewidesnp5",
	  "human370v1c",
	  "human370quadv3c",
	  "human550v3b",
	  "human650v3a",
	  "human610quadv1b",
	  "human660quadv1a",
	  "human1mduov3b")
}

isValidCdfName <- function(cdfName){
	chipList <- validCdfNames()
	result <- cdfName %in% chipList	
	if(!(result)){
		warning("cdfName must be one of the following: ",
			chipList)
	}
	return(result)
}

initializeBigMatrix <- function(dns, vmode="integer"){
##initializeBigMatrix <- function(nr, nc, vmode="integer"){
	nr <- length(dns[[1]])
	nc <- length(dns[[2]])
	if(isPackageLoaded("ff")){
	if(prod(nr, nc) > 2^31){
		##Need multiple matrices
		## -- use ffdf
			
		## How many samples per ff object
		S <- floor(2^31/nr - 1)

		## How many ff objects
		L <- ceiling(nc/S)

		results <- vector("list", L)
		##resultsB <- vector("list", L)			
		for(i in 1:(L-1)){  ## the Lth object may have fewer than nc columns
			results[[i]] <- ff(dim=c(nr, S),
					   vmode=vmode,
					   finalizer="close",
					   overwrite=TRUE)
		}
		##the Lth element
		leftOver <- nc - ((L-1)*S)
		results[[L]] <- ff(dim=c(nr, leftOver),
				   vmode=vmode,
				   finalizer="close",
				   overwrite=TRUE)
		resultsff <- do.call(ffdf, results)
		##dimnames(resultsff) <- dns
	} else {
		resultsff <- ff(dim=c(nr, nc),
				vmode=vmode,
				finalizer="close",
				overwrite=TRUE)
##				dimnames=dns)
	}
	resultsff[,] <- NA
}  else resultsff <- matrix(NA, nr, nc)
	return(resultsff)
}

paramNames <- function(){
	c("tau2A", "tau2B", "sig2A", "sig2B",
	  "nuA", "nuB", "nuA.se", "nuB.se", "phiA", "phiB", "phiA2", "phiB2",
	  "phiA.se", "phiB.se", "corr", "corrA.BB", "corrB.AA")
}

initializeParamObject <- function(dimnames){
	nr <- length(dimnames[[1]])
	nc <- length(dimnames[[2]])		
	ll <- vector("list", 17)
	if(isPackageLoaded("ff")){
		for(i in 1:17) ll[[i]] <- ff(dim=c(nr,nc), vmode="double", finalizer="close", dimnames=dimnames, overwrite=TRUE)
		names(ll) <- paramNames()
		ll <- do.call(ffdf, ll)
	} else {
		for(i in 1:17) ll[[i]] <- matrix(NA, nr, nc, dimnames=dimnames)
		names(ll) <- paramNames()
	}
	return(ll)
}

whichPlatform <- function(cdfName){
	index <- grep("genomewidesnp", cdfName)
	if(length(index) > 0){
		platform <- "Affymetrix"
	} else{
		index <- grep("human", cdfName)
		platform <- "Illumina"
	}
	return(platform)
}

updateParams <- function(cnParams, object, row.index, batch){
	labels.asis <- fvarLabels(object)[4:20]
	B <- batch
	batch <- paste("_", batch, sep="")
	labels <- as.character(sapply(labels.asis, function(x, batch) strsplit(x, batch)[[1]][1], batch=batch))
	labels <- gsub("X", "2", labels)
	if(!isPackageLoaded("ff")){
		##batch <- paste("_", strsplit(labels.asis[1], batch), sep="")
		##labels <- gsub("_", "\\.", labels.asis)
		##labels <- gsub(paste("\\.", batch, sep=""), "", labels)
		j <- match(B, colnames(cnParams[[1]]))
		for(i in seq(along=labels)){
			ii <- match(labels[i], names(cnParams))
			jj <- match(labels.asis[i], fvarLabels(object))
			cnParams[[ii]][, j] <- fData(object)[, jj]
		}
	} else {
		##labels <- as.character(sapply(labels.asis, function(x, batch) strsplit(x, batch)[[1]][1], batch=batch))		
		##labels <- gsub("_", "\\.", labels.asis)
		##labels <- gsub("X", "2", labels)		
		col.index <- match(labels, colnames(cnParams))
		cnParams[row.index, col.index] <- fData(object)[, 4:20]
	}
	return(cnParams)
}

##			tau2A[row.index, i] <- getParam(cnSet, "tau2A", unique(batch)[i])
##			tau2B[row.index, i] <- getParam(cnSet, "tau2B", unique(batch)[i])
##			sig2A[row.index, i] <- getParam(cnSet, "sig2A", unique(batch)[i])
##			sig2B[row.index, i] <- getParam(cnSet, "sig2B", unique(batch)[i])
##			nuA[row.index, i] <- getParam(cnSet, "nuA", unique(batch)[i])
##			nuA.se[row.index, i] <- getParam(cnSet, "nuA.se", unique(batch)[i])					
##			nuB[row.index, i] <- getParam(cnSet, "nuB", unique(batch)[i])
##			nuB.se[row.index, i] <- getParam(cnSet, "nuB.se", unique(batch)[i])										
##			phiA[row.index, i] <- getParam(cnSet, "phiA", unique(batch)[i])
##			phiA.se[row.index, i] <- getParam(cnSet, "phiA.se", unique(batch)[i])																				
##			phiB[row.index, i] <- getParam(cnSet, "phiB", unique(batch)[i])
##			phiB.se[row.index, i] <- getParam(cnSet, "phiB.se", unique(batch)[i])															
##			corr[row.index, i] <- getParam(cnSet, "corr", unique(batch)[i])
##			corrA.BB[row.index, i] <- getParam(cnSet, "corrA.BB", unique(batch)[i])
##			corrB.AA[row.index, i] <- getParam(cnSet, "corrB.AA", unique(batch)[i])
##			if(CHR==23){
##				phiA2[row.index, i] <- getParam(cnSet, "phiA2", unique(batch)[i])
##				phiB2[row.index, i] <- getParam(cnSet, "phiB2", unique(batch)[i])
##			}
##			if(!isPackageLoaded("ff")){
##				save(cnSet, file=paste(cnFile, "_", PLATE, "_", CHR, ".rda", sep=""))				
##				cnParams <- list(tau2A=tau2A,
##						 tau2B=tau2B,
##						 sig2A=sig2A,
##						 sig2B=sig2B,
##						 nuA=nuA,
##						 nuB=nuB,
##						 nuA.se=nuA.se,
##						 nuB.se=nuB.se,
##						 phiA=phiA,
##						 phiA2=phiA2,
##						 phiB=phiB,
##						 phiB2=phiB2,
##						 phiA.se=phiA.se,
##						 phiB.se=phiB.se,
##						 corr=corr,
##						 corrA.BB=corrA.BB,
##						 corrB.AA=corrB.AA)
##				save(cnParams, file=paste(outfile, "cnParams_", PLATE, "_", CHR, ".rda", sep=""))
##			}

##parameterMatrix <- function(object){
##	CHR <- unique(chromosome(object))
##	tau2A[row.index, i] <- getParam(cnSet, "tau2A", unique(batch)[i])
##	tau2B[row.index, i] <- getParam(cnSet, "tau2B", unique(batch)[i])
##	sig2A[row.index, i] <- getParam(cnSet, "sig2A", unique(batch)[i])
##	sig2B[row.index, i] <- getParam(cnSet, "sig2B", unique(batch)[i])
##	nuA[row.index, i] <- getParam(cnSet, "nuA", unique(batch)[i])
##	nuA.se[row.index, i] <- getParam(cnSet, "nuA.se", unique(batch)[i])					
##	nuB[row.index, i] <- getParam(cnSet, "nuB", unique(batch)[i])
##	nuB.se[row.index, i] <- getParam(cnSet, "nuB.se", unique(batch)[i])										
##	phiA[row.index, i] <- getParam(cnSet, "phiA", unique(batch)[i])
##	phiA.se[row.index, i] <- getParam(cnSet, "phiA.se", unique(batch)[i])																				
##	phiB[row.index, i] <- getParam(cnSet, "phiB", unique(batch)[i])
##	phiB.se[row.index, i] <- getParam(cnSet, "phiB.se", unique(batch)[i])															
##	corr[row.index, i] <- getParam(cnSet, "corr", unique(batch)[i])
##	corrA.BB[row.index, i] <- getParam(cnSet, "corrA.BB", unique(batch)[i])
##	corrB.AA[row.index, i] <- getParam(cnSet, "corrB.AA", unique(batch)[i])
##	CA[row.index, sample.index] <- cnSet@assayData[["CA"]]
##	CB[row.index, sample.index] <- cnSet@assayData[["CB"]]
##	if(CHR==23){
##		phiA2[row.index, i] <- getParam(cnSet, "phiA2", unique(batch)[i])
##		phiB2[row.index, i] <- getParam(cnSet, "phiB2", unique(batch)[i])
##	}
##	paramFF <- ffdf(tau2A=tau2A,
##			tau2B=tau2B,
##			sig2A=sig2A,
##			sig2B=sig2B,
##			nuA=nuA,
##			nuB=nuB,
##			nuA.se=nuA.se,
##			nuB.se=nuB.se,
##			phiA=phiA,
##			phiA2=phiA2,
##			phiB=phiB,
##			phiB2=phiB2,
##			phiA.se=phiA.se,
##			phiB.se=phiB.se,
##			corr=corr,
##			corrA.BB=corrA.BB,
##			corrB.AA=corrB.AA)
##	return(paramFF)
##}

constructClass <- function(annotation){
	thisclass <- whichClass(annotation)
	obj <- new(thisclass, annotation=annotation)
	crlmmOptions(obj) <- getOptions(obj)
	obj
}

whichClass <- function(cdfName){
	stopifnot(isValidCdfName(cdfName))
	platform <- whichPlatform(cdfName)
	if(isPackageLoaded("ff")){
		thisclass <- paste(platform, "AlleleSet", sep="")
	} else {
		thisclass <- paste(platform, "AlleleSet", sep="")
	}
	return(thisclass)
}
setMethod("annotatedDataFrameFrom", "ff_matrix", Biobase:::annotatedDataFrameFromMatrix)
